flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/ExecutorWrapper.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.kubernetes.kubeclient;
+
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.kubernetes.configuration.KubernetesConfigOptions;
+import org.apache.flink.runtime.util.ExecutorThreadFactory;
+import org.apache.flink.util.ExecutorUtils;
+
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * The help class to get the executor to run asynchronous operations on. It could create a dedicated thread pool or
+ * reuse an existing one.
+ */
+public class ExecutorWrapper {
+
+	private ExecutorService internalExecutorService;
+	private Executor executor;
+
+	private ExecutorWrapper(Configuration flinkConfig) {
+		internalExecutorService = Executors.newFixedThreadPool(
+			flinkConfig.getInteger(KubernetesConfigOptions.CLIENT_ASYNC_THREAD_POOL_SIZE),
+			new ExecutorThreadFactory("FlinkKubeClient-IO"));
+	}
+
+	private ExecutorWrapper(Executor executor) {
```

Comment:
The reason why i use `ExecutorWrapper` is to provide factory method to create a dedicated thread pool or reuse an existing one. However, i think this could also be done to make the argument of `Fabric8FlinkKubeClient` `Nullable`. I will go in this way.

Reply:
It does not have to be `@Nullable`.  It is better to create all dependencies outside of the class in its factory. The `Executors.newFixedThreadPool` can be created in `KubeClientFactory`.

Reply:
we can also shutdown any executor in `Fabric8FlinkKubeClient#close`, not sure why we need two fields `internalExecutorService` and `kubeClientExecutorService`

Reply:
In my opinion, who created the executor then it is responsible for its lifecycle. That's why i use the `internalExecutorService`. So if create the executor service outside of `Fabric8FlinkKubeClient` and shutdown it in `Fabric8FlinkKubeClient#close`. It may cause some problems since some other components are still using it.

Of cause, we could remove the `@Nullable` for `kubeClientExecutorService`, then we may need to add another argument to control whether need to shutdown it.
WDYT?

Reply:
You could write in comments that you hand over the lifecycle of the executor to the client. I do not see a big problem with that, it is created in its factory or tests, only for the client.
Alternatively, you could pass a factory for the `ExecutorService` to the client:
`Supplier<ExecutorService> asyncExecutorFactory`
then it is created and shutdown within the client.
It seems to me more clear approach with less ifs and vars.

Anyways, once we have the general purpose executor in RpcEndpoint [(FLINK-17024)](https://issues.apache.org/jira/browse/FLINK-17024), it can be refactored.

Reply:
Make sense. I will pass a executor service factory to `Fabric8FlinkKubeClient` and shutdown it in the `Fabric8FlinkKubeClient#close`.